### ISSUE TITLE
[$250] Messages - Latest message instead of marked as unread after clicking 'New messages'

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#44647]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.3-2
**Reproducible in staging?:** Y
**Reproducible in production?:** Y 
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** test.iris.santos+17062024@gmail.com
**Issue reported by:** Applause - internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. Navigate to a 1:1 conversation that has a large history of messages.
3. Scroll the conversation a little and hover over a message from the other user and mark the message as "Unread"
4. Scroll further up the conversation
5. Click on "New Messages" button

## Expected Result:
User is scrolled to the message marked as unread after clicking the 'New messages' button

## Actual Result:
User is scrolled to the most recent message after clicking the 'New messages' button

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/1770c5ba-2208-4f38-98a0-af76a3a7eca1

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c686d04b6354397a</li>
        <li>Upwork Job ID: 1808416574319645071</li>
        <li>Last Price Increase: 2024-11-13</li>
<li>Automatic offers: </li>
<ul>
<li>rojiphil | Reviewer | 104915994</li>
<li>nkdengineer | Contributor | 104915997</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @rojiphil</details>

**Comment by Sparth19** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Clicking ‘New Messages’ scrolls to the Latest Message Instead of the First Unread Message


### What is the root cause of that problem?

According to the [`scrollToBottomAndMarkReportAsRead`](https://github.com/Expensify/App/blob/d489ba44a1d4c951de7a405caa452f33b86ac70f/src/pages/home/report/ReportActionsList.tsx#L406-L415) method, it always scrolls to the bottom.

### What changes do you think we should make in order to solve the problem?

- We can update the `scrollToBottomAndMarkReportAsRead` method to use `scrollToIndex` and provide the `firstUnreadMessageIndex`, ensuring it navigates to the first unread message instead of always scrolling to the bottom.

so change

`reportScrollManager.scrollToBottom(); 
`

to

```
if (!currentUnreadMarker) return;
const unreadMessageIndex = sortedVisibleReportActions.findIndex((action) => action.reportActionID === currentUnreadMarker);
if (unreadMessageIndex !== -1) {
    reportScrollManager?.scrollToIndex(unreadMessageIndex, false);
}
```

### What alternative solutions did you explore? (Optional)

**Comment by dylanexpensify** [comment]
Reviewed!

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
User is scrolled to the most recent message after clicking the 'New messages' button

### What is the root cause of that problem?
In `scrollToBottomAndMarkReportAsRead` [here](https://github.com/Expensify/App/blob/4622e19c1da5137bc7b227f5b65bf3d83f96a909/src/pages/home/report/ReportActionsList.tsx#L412), we always scroll to bottom.

### What changes do you think we should make in order to solve the problem?

1. We need to update [this](https://github.com/Expensify/App/blob/4622e19c1da5137bc7b227f5b65bf3d83f96a909/src/pages/home/report/ReportActionsList.tsx#L412) so that it will scroll to the unread message instead. When scrolling to the unread message, we need to make sure the unread message will show on top so the user can see as much of the unread messages as possible

First update [this](https://github.com/Expensify/App/blob/main/src/hooks/useReportScrollManager/index.ts#L11-L20) to accept an optional `viewPosition` param
```
const scrollToIndex = useCallback(
    (index: number, isEditing?: boolean, viewPosition?: number) => {
        if (!flatListRef?.current || isEditing) {
            return;
        }

        flatListRef.current.scrollToIndex({index, animated: true, viewPosition});
    },
    [flatListRef],
);
```
Next remove [this line](https://github.com/Expensify/App/blob/8e2831c4c9cc3844c5ac9933787ad5a01bda82f1/src/pages/home/report/ReportActionsList.tsx#L412)

And add [here](https://github.com/Expensify/App/blob/8e2831c4c9cc3844c5ac9933787ad5a01bda82f1/src/pages/home/report/ReportActionsList.tsx#L415)
```
const unreadReportActionIndex = sortedVisibleReportActions.findIndex((action) => action.reportActionID === currentUnreadMarker);
if (unreadReportActionIndex !== -1) {
    reportScrollManager?.scrollToIndex(unreadReportActionIndex, undefined, 0.9);
} else {
   reportScrollManager.scrollToBottom();
}
```
We'll scroll to bottom as usual if the `unreadReportActionIndex` couldn't be found. The `0.9` `viewPosition` can be adjusted, I think `0.9` makes sense because not only does it make sure the user can see all the unread messages, it also makes sure the `New` marker above the unread message still shows fully

2. After the above fix, we face another issue that when we click the `New messages` it will scroll to the unread message. But the `New messages` will still be visible and if we click it again, nothing happens. This is due to in the second attempt, `New messages` will still attempt to scroll to the unread message and since the list was already scrolled there, it's unresponsive and looks broken to the user.

I think if there's already the `New` marker above the unread message, we don't need to show the `New messages` CTA as the user is already seeing there's new message.

So we should store the `New` marker above the unread message/the unread message report action item in a ref, and in `ReportActionsList`, use that ref to check if the `New` marker is visible in the screen (by getting it's location, scroll offset, or use bounding rect). If it's visible, do not show the `New messages` CTA [here](https://github.com/Expensify/App/blob/8e2831c4c9cc3844c5ac9933787ad5a01bda82f1/src/pages/home/report/ReportActionsList.tsx#L679)

### What alternative solutions did you explore? (Optional)

**Comment by dylanexpensify** [comment]
bump @rojiphil

**Comment by rojiphil** [comment]
Reviewing today

**Comment by rojiphil** [comment]
@dylanexpensify I think it is intentional to scroll to the bottom when `New Messages` is clicked. And for the `unread` messages, it becomes `read` when the user navigates away and comes back to the same report. This has been so for the last one year. So, I am not sure if we have a real issue to fix. Can we please reconfirm the expected behavior here?

**Comment by nkdengineer** [comment]
> @dylanexpensify I think it is intentional to scroll to the bottom when New Messages is clicked.

@rojiphil I don't think this is good in terms of the user's experience. They will miss all the unread messages in between (if there're a lot of unread messages and we only scroll to the bottom, the user will most likely only see the last page of messages)

**Comment by rojiphil** [comment]
> I don't think this is good in terms of the user's experience.

Well! That's debatable as the `unread` messages seem to be designed the way it is now. 
Let me bring in @Expensify/design and get their views on my comment [here](https://github.com/Expensify/App/issues/44647#issuecomment-2226101274)

**Comment by dannymcclain** [comment]
🤔 hmm, definitely want to hear the other designer's thoughts, but I think I would assume that clicking on the `New messages` pill would take you to the message with the `New` tag on it.

In a chat with no recent new messages (so nothing unread at the "bottom" of the chat) it looks like the `New messages` pill doesn't show up until you mark _some_ message as unread. Which makes me feel like they should be related somehow. But I don't know the ins and outs of how this works the way it does or why.

**Comment by dubielzyk-expensify** [comment]
Yeah, I agree with Danny that tapping `New messages` should go to the `New` tag

**Comment by shawnborton** [comment]
Yup, +1 to both of yas.

**Comment by rojiphil** [comment]
Thanks design team for the inputs.

@dylanexpensify Based on the comment [here](https://github.com/Expensify/App/issues/44647#issuecomment-2226101274) and the feedback from the design team, the current behavior is expected. 
I think we can close this issue.

**Comment by dannymcclain** [comment]
🤔 I think the design team is saying that the current behavior is _not_ expected and that the "Expected behavior" outlined in the issue description is correct, but the product does not match that behavior. Am I missing something?

**Comment by rojiphil** [comment]
> current behavior is _not_ expected and that the "Expected behavior" outlined in the issue description is correct

@dannymcclain Oh! Let's get this sorted then.  
1. The expected result mentioned in OP is `User is scrolled to the message marked as unread after clicking the 'New messages' button`. 
2. But as per design team's comments `tapping New messages should go to the New tag`. 
3. And the current behavior as per the actual result in OP is: `User is scrolled to the most recent message after clicking the 'New messages' button`

As I understand, current behavior and the design team's comments are aligned. Or am I missing something here?

**Comment by nkdengineer** [comment]
@rojiphil I believe `tapping New messages should go to the New tag` means "go to the first unread message" (that has the `New` part), not "go to the most recent message"

**Comment by dannymcclain** [comment]
>I believe tapping New messages should go to the New tag means "go to the first unread message" (that has the New part), not "go to the most recent message"

Correct! We're saying that the `New messages` pill should scroll you to the message that has the green `New` label on it.

So in this example below, there are many more messages after the one I've marked as unread/new, but clicking the green New messages pill should take me to this message.

![CleanShot 2024-07-23 at 11 23 36@2x](https://github.com/user-attachments/assets/9b115bc1-91aa-44f7-a77a-67f7c0117e1e)

**Comment by rojiphil** [comment]
> tapping New messages should go to the New tag means "go to the first unread message" (that has the New part)

> there are many more messages after the one I've marked as unread/new

Ah! Thanks @dannymcclain, @nkdengineer. Now I can see where I got stuck. 
When we mark an already read message as `unread`, we consider this as `new`. 
And I think the words `unread` and `new` tricked me as it seemed strange to me to call a once `read` message as `new` at a later point of time by marking it as `unread`. 
Maybe we should have set the menu option as `Mark as new`

**Comment by rojiphil** [comment]
And, regarding the proposals, I will re-review my day tomorrow and update.

**Comment by dubielzyk-expensify** [comment]
Yep, absolutely agree with what @dannymcclain is saying

**Comment by rojiphil** [comment]
Thanks all for your proposals.
@nkdengineer [proposal](https://github.com/Expensify/App/issues/44647#issuecomment-2205862683) LGTM as the proposal has the additional advantage of a) using `viewPosition` to adjust the position in the viewable area b) hiding the `New messages` floating message even when the current unread marker is visible in the viewable area.  
🎀👀🎀 C+ reviewed

**Comment by rojiphil** [comment]
> Solution on the selected proposal not working properly if there are new messages after marked as unread:

@nyomanjyotisa I am not sure if I understood you correctly here. Can you please share the test steps for this and the expected result and actual result? And also please mention the actual result observed in selected proposal and the specific code that makes your proposal better.

**Comment by rojiphil** [comment]
> Specific code on my proposal that make it work on this case: [`scrollToItem`](https://github.com/Expensify/App/issues/44647#:~:text=reportScrollManager.scrollToItem((currentUnreadMarkerItem))%3B)

@nyomanjyotisa Thanks for sharing the details but I am still trying to understand the difference. 
Your proposal sets the current unread marker as `currentUnreadMarkerItem` and uses it whereas the selected proposal fetches the same from `sortedVisibleReportActions`. What's the difference here or am I missing something here?

**Comment by nkdengineer** [comment]
> @nkdengineer Can you please investigate why this https://github.com/Expensify/App/issues/44647#issuecomment-2250195414 would fail for your proposal?

Sure, I'll give an update on this tomorrow

**Comment by nkdengineer** [comment]
I will post the update soon

**Comment by nkdengineer** [comment]
This is a bit tricky since it seems related to the internal implementation of `FlatList`, I'm still investigating and will try to update soon.

**Comment by dylanexpensify** [comment]
Thank you!

**Comment by stitesExpensify** [comment]
@nkdengineer is still trying to figure this out

**Comment by stitesExpensify** [comment]
not overdue

**Comment by nkdengineer** [comment]
Still investigating, I'm looking into a reasonable alternative we can use that still achieves the desired behavior.

**Comment by nkdengineer** [comment]
I found the root cause! Turns out it's a bug deep inside `react-native-web` library, trying to come up with a solution now.

**Comment by stitesExpensify** [comment]
wow crazy. Looking forward to the solution!

**Comment by nkdengineer** [comment]
Proposal updated with the fix for the issue below
> ### To address [this issue](https://github.com/Expensify/App/issues/44647#issuecomment-2250399641), it doesn't happen on native, only web, because of a bug in `react-native-web`. 

@alitoshmatov Please kindly review, thanks

**Comment by rojiphil** [comment]
> To address [this issue](https://github.com/Expensify/App/issues/44647#issuecomment-2250399641), it doesn't happen on native, only web, because of a bug in `react-native-web`.

@nkdengineer Can you please share a test branch with the proposed changes?

**Comment by nkdengineer** [comment]
@rojiphil Yeah here it is https://github.com/nkdengineer/App/tree/fix/44647

**Comment by stitesExpensify** [comment]
Development is moving along!

**Comment by rojiphil** [comment]
> Yeah here it is https://github.com/nkdengineer/App/tree/fix/44647

Thanks. Review is in progress. Will revert in few days

**Comment by stitesExpensify** [comment]
Review in progress

**Comment by rojiphil** [comment]
> To fix this, we need to update the other frames too if a frame comes that pushes the index and offset in other frames (think the case of new item coming at index 0, then all other frames's indexes will increase and we need to reflect that in the `_frames` storage).

@nkdengineer Thanks for sharing the test PR. While I agree with the analysis i.e. the problem with `index` in `_frames`, the proposed solution looks like a hack to me. Instead of trying to fix the problem only for the case of new actions, we need to solve such that any mismatch between the modified index and stored index (frame.index) is fixed holistically. Otherwise, we will miss some cases. For example, deletion of an `action` in between will also cause the indexes to mismatch, causing the same problem. Can you think of some other solution? Could using `extraData` or `getItemLayout` help solve the problem?

**Comment by stitesExpensify** [comment]
Hi @nkdengineer any thoughts on the above? ^

**Comment by nkdengineer** [comment]
Hey I'll update on that tomorrow 👍

**Comment by nkdengineer** [comment]
> Could using extraData or getItemLayout help solve the problem?

@rojiphil We cannot use `getItemLayout` because report action can be of arbitrary sizes and there is no reliable way to know the size before the item is rendered

We should not use `extraData` because it might lead to rerendering of the whole list when 1 single item change, which we don't want. The current way of rendering only the new item is efficient IMO.

> Instead of trying to fix the problem only for the case of new actions, we need to solve such that any mismatch between the modified index and stored index (frame.index) is fixed holistically.

@rojiphil Yes we can fix it holistically, we just need to expand my solution for those cases, for example in the `1 item deletion` case, we also will know that the item is deleted and can shift the `index` of the remaining items accordingly.

I don't think my solution is a hack because the values in `_frames` after an item is added/deleted are obviously incorrect and has to be fixed one way or another. I think using diffing method like in my solution is correct and efficient because we could always know what changes were made to the item (added/deleted) by comparing the new value to the existing value in `_frames`, it's not much different from how `git` softwares work so the approach is valid IMO.

Let me know what you think.

**Comment by dylanexpensify** [comment]
bump @rojiphil for thoughts

**Comment by stitesExpensify** [comment]
Discussion is ongoing, not overdue

**Comment by nkdengineer** [comment]
@rojiphil Gentle bump on https://github.com/Expensify/App/issues/44647#issuecomment-2334450684
Thanks

**Comment by rojiphil** [comment]
Sorry for the delay here. I will share my feedback today.

**Comment by rojiphil** [comment]
@nkdengineer Yeah it makes sense not to use `getItemLayout` for the reasons mentioned. 
But I still prefer not to change the upstream library unless otherwise absolutely required as it adds further complexity to code management. 
In this case, the upstream library allows us to re-render cells from FE. And so I would expect this to be resolved from FE.
Can you think of a solution to resolve the issue without changing the upstream library?

**Comment by stitesExpensify** [comment]
No update, waiting to see if @nkdengineer (or anyone else) can come up with a solution

**Comment by trjExpensify** [comment]
>But I still prefer not to change the upstream library unless otherwise absolutely required as it adds further complexity to code management.

Can you expand on why? We typically favour fixing issues upstream if that's the root cause for a more complete fix over a "workaround" in ours.

**Comment by rojiphil** [comment]
> > But I still prefer not to change the upstream library unless otherwise absolutely required as it adds further complexity to code management.
> 
> Can you expand on why? We typically favour fixing issues upstream if that's the root cause for a more complete fix over a "workaround" in ours.

As mentioned in my comment [here](https://github.com/Expensify/App/issues/44647#issuecomment-2351083598), the upstream library allows us to re-render cells from FE. And so I would expect this to be resolved from FE. But if it is not possible to re-render cells from FE due to some inherent issue in upstream, only then would I see the need to fix upstream.

**Comment by dylanexpensify** [comment]
@trjExpensify thoughts on @rojiphil's last comment? I agree we'd typically fix the upstream issue if we know it exists to create better overall code, even if a higher lift.

**Comment by trjExpensify** [comment]
>As mentioned in my comment https://github.com/Expensify/App/issues/44647#issuecomment-2351083598, the upstream library allows us to re-render cells from FE. And so I would expect this to be resolved from FE. But if it is not possible to re-render cells from FE due to some inherent issue in upstream, only then would I see the need to fix upstream.

Cool, makes sense if the bug isn't actually upstream. Let's confirm that then.

**Comment by nkdengineer** [comment]
> In this case, the upstream library allows us to re-render cells from FE. And so I would expect this to be resolved from FE.
Can you think of a solution to resolve the issue without changing the upstream library?

I'm exploring this and will try to give an update tomorrow

**Comment by nkdengineer** [comment]
Sorry I haven't been able to finish on this, on my list for tomorrow

**Comment by stitesExpensify** [comment]
Not overdue, waiting for an update from @nkdengineer

**Comment by dylanexpensify** [comment]
Bump @nkdengineer

**Comment by nkdengineer** [comment]
Sorry for the delay everyone 🙏 .

> In this case, the upstream library allows us to re-render cells from FE. And so I would expect this to be resolved from FE.

@rojiphil @trjExpensify  I've looked into it further and I don't think triggering re-render of all cells when a cell is changed is a good idea. If an item is inserted, it's best to render only that item without re-rendering other items in the list since such widespead re-rendering will cause performance problems.

Also, to reiterate, this issue **doesn't happen on native, only web**. So I think doing some kinds of custom manipulation/re-rendering in our code for this particular web use case is platform-specific and thus workaround. If this is not an issue upstream, then why does it work for native but not web? As a library user I'd expect the library to work well in this case without such tweaks.

**Comment by rojiphil** [comment]
> Also, to reiterate, this issue **doesn't happen on native, only web**. So I think doing some kinds of custom manipulation/re-rendering in our code for this particular web use case is platform-specific and thus workaround.

@nkdengineer Thanks for the response. Now, that's very much convincing to me that there is a fix needed in the upstream library. 
But I am still not convinced of the proposed solution though. Iterating through all the elements and fixing the index whenever cell mounts/unmounts seems like a workaround to me.
Can you think of a solution within upstream library where we fix the indexes when `data` item count changes? This would be much neater way to do.

**Comment by nkdengineer** [comment]
Updating today 🙏

**Comment by nkdengineer** [comment]
> Can you think of a solution within upstream library where we fix the indexes when data item count changes? This would be much neater way to do.

@rojiphil There'd be case where there's a new message but another message was deleted, and both updates are reflected to data at the same time. In that case `data` item count will remain the same, and your suggested approach will not detect that and the indexes will remain incorrect.

**Comment by stitesExpensify** [comment]
Not overdue

**Comment by dylanexpensify** [comment]
@rojiphil thoughts on above?

**Comment by dylanexpensify** [comment]
Bump @rojiphil

**Comment by dylanexpensify** [comment]
@nkdengineer curious what you think?

**Comment by nkdengineer** [comment]
I'll give an update within a day

**Comment by nkdengineer** [comment]
> But wouldn't we know when data changes? And when data changes, can we not figure out that the indexes are out of sync for the first or last item in the data and only then reset the indexes? Do you think this makes sense?

This seems to be a feasible approach, it's similar to my proposal but relies on `data` changes instead of layout event. I'll need a bit more time to adjust my implementation in the proposal to do that and test it.

**Comment by dylanexpensify** [comment]
@rojiphil thoughts?

**Comment by rojiphil** [comment]
> @rojiphil thoughts?

Looks like we are pretty close to get this going. Just waiting for the proposal updation as mentioned [here](https://github.com/Expensify/App/issues/44647#issuecomment-2411264072)

**Comment by dylanexpensify** [comment]
@rojiphil thoughts?

**Comment by rojiphil** [comment]
> And when `data` changes, can we not figure out that the indexes are out of sync for the first or last item in the `data` and only then reset the indexes?

> I've tried to follow this approach in the implementation but unfortunately I don't think it can work. The `frames` does not store just the indexes, it also store layout-related information like the `offset` and `length` ([here](https://github.com/necolas/react-native-web/blob/54c14d64dabd175e8055e1dc92e9196c821f9b7d/packages/react-native-web/src/vendor/react-native/VirtualizedList/index.js#L1288-L1289)).> So it's not possible to trigger the `frames` reset from the data changes because data changes don't have any layout information.

@nkdengineer I may be missing something, but I still don't understand why we need the layout information here. As I understand it, we only have to update the current `_frames` indexes. Did we run into any problems with this?

**Comment by dylanexpensify** [comment]
Hey @puneetlath! I'm heading out on parental leave so reassigning this! TY! 🙇‍♂️

**Comment by nkdengineer** [comment]
> @nkdengineer I may be missing something, but I still don't understand why we need the layout information here. As I understand it, we only have to update the current _frames indexes.

@rojiphil We need the `offset` and `length` to be correct too since it's part of the `_frames` and they are the useful fields of the `_frames` to be used ([here](https://github.com/necolas/react-native-web/blob/54c14d64dabd175e8055e1dc92e9196c821f9b7d/packages/react-native-web/src/vendor/react-native/VirtualizedList/index.js#L1733), [here](https://github.com/necolas/react-native-web/blob/54c14d64dabd175e8055e1dc92e9196c821f9b7d/packages/react-native-web/src/vendor/react-native/VirtualizedList/index.js#L1744), [here](https://github.com/necolas/react-native-web/blob/54c14d64dabd175e8055e1dc92e9196c821f9b7d/packages/react-native-web/src/vendor/react-native/VirtualizedList/index.js#L251C11-L251C16) are examples).

So in short, we need to make sure the full `_frames` object is correct, not just the `_frames` indexes. If the indexes of each `_frames` item are correct but the `offset` and `length` of those items are wrong, the calculation will still be wrong.

**Comment by rojiphil** [comment]
> So in short, we need to make sure the full `_frames` object is correct, not just the `_frames` indexes. If the indexes of each `_frames` item are correct but the `offset` and `length` of those items are wrong, the calculation will still be wrong. In the sample code change in [my proposal](https://github.com/Expensify/App/issues/44647#issuecomment-2205862683), all the fields in `_frames` are updated to correct values.

@nkdengineer Thanks for the clarifications. 
I think the changes in `onCellLayout` are fine, but what is the reasoning behind the check for only `index == 0` in the proposal?   Why not for all indexes i.e. `index < Object.keys(this._frames).length`?
Also, what’s the significance of checking for `doesFrameWithTheSameIndexExist`?
And please update the proposal to handle the case for deleted report actions.

**Comment by stitesExpensify** [comment]
@nkdengineer any update?

**Comment by nkdengineer** [comment]
Thanks for your patience, I needed some time to address [all the concerns](https://github.com/Expensify/App/issues/44647#issuecomment-2427459634)

> but what is the reasoning behind the check for only index == 0 in the proposal? Why not for all indexes i.e. index < Object.keys(this._frames).length?

@rojiphil Yes there's no need to limit to only `index === 0`, I added that initially to illustrate how it works for this particular case, I've updated [the proposal](https://github.com/Expensify/App/issues/44647#issuecomment-2205862683) with `index < Object.keys(this._frames).length`

> Also, what’s the significance of checking for doesFrameWithTheSameIndexExist?

We only have this incorrect index case if there's a new item taking that `index`, and there's an existing frame with that same `index` (which is stale). The `doesFrameWithTheSameIndexExist` is to check that. If `doesFrameWithTheSameIndexExist` is `false`, that means the updated item layout does not displace any other cells and there's no need for the index adjustment for the whole `_frames` list.

> And please update the proposal to handle the case for deleted report actions.

[Proposal updated](https://github.com/Expensify/App/issues/44647#issuecomment-2205862683) to add this

**Comment by rojiphil** [comment]
> We only have this incorrect index case if there's a new item taking that `index`, and there's an existing frame with that same `index` (which is stale). The `doesFrameWithTheSameIndexExist` is to check that. If `doesFrameWithTheSameIndexExist` is `false`, that means the updated item layout does not displace any other cells and there's no need for the index adjustment for the whole `_frames` list.

@nkdengineer Thanks for the update. I don’t think we need to check for  `doesFrameWithTheSameIndexExist` here as the parent `if` condition indicates that there is a need for an adjustment. Also, since we are adjusting the indexes during deletion of report actions, relying on the pre-existing index during addition of an existing frame does not seem to be the right thing to do especially since the deleted report action’s index is anyway taken by someone else. So this would mean that even the existing frame would need an adjustment of index during addition. One possible way to overcome this is to reset the index to `-1` during deletion which will trigger reassignment on addition later. Overall I think we are in the right direction. So, let us explore this more in the PR stage.

**Comment by rojiphil** [comment]
@nkdengineer [proposal](https://github.com/Expensify/App/issues/44647#issuecomment-2205862683) to reassign the indexes during addition and removal of report actions LGTM. 
🎀👀🎀 C+ reviewed
Cc @puneetlath @stitesExpensify

**Comment by puneetlath** [comment]
bump @stitesExpensify

**Comment by puneetlath** [comment]
@nkdengineer @rojiphil what's the status here?

**Comment by rojiphil** [comment]
@puneetlath The PR is actively been reviewed with the latest being [here](https://github.com/Expensify/App/pull/52689#issuecomment-2530806838) and [here](https://github.com/Expensify/App/pull/52689#issuecomment-2533668350)

**Comment by nkdengineer** [comment]
@puneetlath  Based on [here](https://github.com/Expensify/App/pull/52689#issuecomment-3916418081), we can open this for proposal now.

**Comment by dmytrodyakonov98-tech** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When a user has marked an older message as unread, scrolls further back in the conversation, and presses the **New messages** floating pill, the list scrolls to the newest message instead of the message that owns the unread marker.

### What is the root cause of that problem?
The current code already computes the correct unread target in `useUnreadMarker()` and passes both `unreadMarkerReportActionID` and `unreadMarkerReportActionIndex` into `useReportActionsScroll()`.

However, `scrollToBottomAndMarkReportAsRead()` does not use that index. When the newest actions are already loaded, it always calls:

```ts
reportScrollManager.scrollToBottom();
markNewestActionAsRead();
```

So even when `unreadMarkerReportActionIndex >= 0`, clicking **New messages** explicitly targets index `0` of the inverted list (the newest action) rather than the unread-marker action.

This is different from the older implementation discussed in this issue: the current scroll manager already exposes `scrollToIndex()`, so we do not need to patch VirtualizedList frame bookkeeping to select the correct target.

### What changes do you think we should make in order to solve the problem?
Use the existing `unreadMarkerReportActionIndex` as the preferred target in `scrollToBottomAndMarkReportAsRead()` when the newest action is already present:

```ts
if (unreadMarkerReportActionIndex >= 0) {
    reportScrollManager.scrollToIndex(unreadMarkerReportActionIndex);
} else {
    reportScrollManager.scrollToBottom();
}
markNewestActionAsRead();
```

Keep the existing `!hasNewestReportAction` branch unchanged. In that case the current visible slice may not contain the newest actions/target yet, so the existing navigate/open-report/load-newer behavior should remain the fallback.

This keeps the current mark-as-read behavior intact; it only changes the scroll destination when a concrete unread-marker action already exists in the same `sortedVisibleReportActions` index domain used by the list.

### Tests
Update `tests/unit/useReportActionsScrollTest.tsx` to cover:

1. `unreadMarkerReportActionIndex >= 0` + newest action loaded → calls `scrollToIndex(unreadMarkerReportActionIndex)`, does **not** call `scrollToBottom()`, and still calls `markNewestActionAsRead()` once.
2. No unread marker (`-1`) + newest action loaded → preserves the existing `scrollToBottom()` + mark-as-read behavior.
3. Newest action not loaded → preserves the existing navigation/open-report/scroll-to-bottom path.

I can implement this against current `main` if this direction is accepted.
